tokio/src/util/sharded_list.rs

```diff
@@ -0,0 +1,160 @@
+    /// Creates a new and empty sharded linked list under specified size
+    pub(crate) fn new(sharded_size: usize) -> Self {
+        // Find power-of-two sizes best matching arguments
+        let mut size = 1;
+        while size < sharded_size {
+            size <<= 1;
+        }
+        let shard_mask = size - 1;
+
+        let mut lists = Vec::with_capacity(size as usize);
+        for _ in 0..size {
+            lists.push(Mutex::new(LinkedList::<L, T>::new()))
+        }
+        Self {
+            lists: lists.into_boxed_slice(),
+            count: AtomicUsize::new(0),
+            shard_mask,
+        }
+    }
+}
+
+/// Used to get the lock of shard
+pub(crate) struct ShardGuard<'a, L, T> {
+    lock: MutexGuard<'a, LinkedList<L, T>>,
+    count: &'a AtomicUsize,
+    id: usize,
+}
+
+impl<L: ShardedListItem> ShardedList<L, L::Target> {
+    #[allow(dead_code)]
+    /// Push a value to this shard.
+    pub(crate) fn push(&self, val: L::Handle) {
+        let shard = self.lock_shard(&val);
+        shard.push(val);
+        self.count.fetch_add(1, Ordering::Relaxed);
+    }
+
+    /// Removes the last element from a list specified by shard_id and returns it, or None if it is
+    /// empty.
+    pub(crate) fn pop_back(&self, shard_id: usize) -> Option<L::Handle> {
+        let mut lock = self.shard_inner(shard_id);
+        let node = lock.pop_back();
+        if node.is_some() {
+            self.count.fetch_sub(1, Ordering::Relaxed);
+        }
+        node
+    }
+
+    /// Removes the specified node from the list
+    ///
+    /// # Safety
+    ///
+    /// The caller **must** ensure that exactly one of the following is true:
+    /// - `node` is currently contained by `self`,
+    /// - `node` is not contained by any list,
+    /// - `node` is currently contained by some other `GuardedLinkedList`.
+    pub(crate) unsafe fn remove(&self, node: NonNull<L::Target>) -> Option<L::Handle> {
+        let id = L::get_shared_id(node);
+        let mut lock = self.shard_inner(id);
+        let node = lock.remove(node);
+        if node.is_some() {
+            self.count.fetch_sub(1, Ordering::Relaxed);
+        }
+        node
+    }
+
+    /// Gets the lock of ShardedList, makes us have the write permission.
+    pub(crate) fn lock_shard(&self, val: &L::Handle) -> ShardGuard<'_, L, L::Target> {
+        let id = unsafe { L::get_shared_id(L::as_raw(val)) };
+        ShardGuard {
+            lock: self.shard_inner(id),
+            count: &self.count,
+            id,
+        }
+    }
+
+    /// Gets the count of elements in this list.
+    pub(crate) fn len(&self) -> usize {
+        self.count.load(Ordering::Relaxed)
+    }
+
+    /// Returns whether the linked list does not contain any node
+    pub(crate) fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Gets the shard size of this SharedList.
+    /// Used to help us to decide the parameter `shard_id`` of the `pop_back` method.
```

Comment:
Generally, you have an empty after the first line of documentation.
```suggestion
    /// Gets the shard size of this SharedList.
    ///
    /// Used to help us to decide the parameter `shard_id`` of the `pop_back` method.
```